### DISCUSSION TITLE
Web-based (PWA) version of OpenSC ?

### DISCUSSION THREAD

**Discussion by VVWTest** [OpenSC/OpenSC#2908] [General]
Hi all
I would like to know if there is any interest (or existing efforts) on building a web-based version of OpenSC which can be used from a browser (like a Chrome extension or firefod add-on) and thus be OS-agnostic.
Just checking - it would be interesting to know if the possibility exists and if not, what are the challenges.
Thanks for a great solution !
Thanks
bby

**Comment by frankmorgner** [comment]
That's not possible at the moment, because there is no platform agnostic way of accessing a hardware token. W3C offers the Cryptography API, but this only exposes what the browser implements/supports. Hardware token drivers are not a part of any browser.

You could wrap the platform specific code into a platform specific browser extension, but that means that you'd have to maintain a couple of versions of your extensions. The easiest way out, would be to use some native application which forwards the requests to the token to the browser, e.g. by using a websocket. For example, you could use some existing program to forward the hardware token via TCP/IP (e.g. [`pcsc-relay`](http://frankmorgner.github.io/vsmartcard/pcsc-relay/README.html) or [`vicc --type=relay`](http://frankmorgner.github.io/vsmartcard/virtualsmartcard/README.html)), map this to a websocket (e.g. with [websockify](https://github.com/novnc/websockify)) and then use this websocket for I/O with your [WebAssembly](https://webassembly.org/) compiled version of OpenSC in the PWA.

Since you always need some platform specific code, you can also ship a native app that's doing *everything* you need.

**Comment by VVWTest** [comment]
Thanks a lot, frankmorgner!
I'd been checking on the possibilities of using Isolated Web Apps -
https://github.com/WICG/isolated-web-apps/blob/main/README.md

**Reply by frankmorgner** [reply]
From the proposal, I cannot see how native code can be added to an isolated web app (as accessing a smart card is not possible through any Browser API). So I guess, this is still not possible without having a native application that is running seperately from the browser.
